### DISCUSSION TITLE
Modal: dynamically change `aria-hidden` based on modal state

### DISCUSSION THREAD

**Discussion by tinymachine** [incluud/accessible-astro-components#20] [General]
Closed modals don't have `aria-hidden` set to false, so screen readers (like VoiceOver on macOS and iOS) will still read the contents of closed modals. This should probably be defaulted to true, and changed to false in the `openModal()` method, and reset to true in `closeModal()`.

**Comment by markteekman** [comment]
Good find @tinymachine, thank you! I'll add it to the list 🙂

**Comment by tinymachine** [comment]
My pleasure, @markteekman!

**Comment by markteekman** [comment]
@tinymachine tested it myself and read into it a little more but it should work fine as it is! The modal has it's `visibility` at `hidden` and that should be enough: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden.

**Comment by tinymachine** [comment]
Thanks for explaining, @markteekman! That behavior makes perfect sense, and works correctly for me on your [Modal demo page](https://components.accessible-astro.dev/modal/). Weirdly, I'm experiencing an issue where embedded Vimeo videos in modals are being read by VoiceOver on iOS 15.4.1 even when the modals are closed (i.e. `visibility: hidden`), but adding `aria-hidden="true"` to the modal element fixes it. Not sure if a bug in my code or maybe in Safari/VoiceOver? I'll troubleshoot and let you know if I sort it out. Thanks again!

**Comment by markteekman** [comment]
@tinymachine it got me thinking, could it have anything to do with the embedded Vimeo video being an `iframe`? And thus it being handled differently as a child of a parent with `visibility: hidden`?

**Comment by tinymachine** [comment]
I think you're exactly right — this is not at all an issue with your code, or with Vimeo. It looks like there's a Safari bug whereby VoiceOver reads the contents of `iframe`s even when the parent has `visibility: hidden`. Here's a screen capture of VoiceOver on iOS 15.4.1 reading a [test page][tp] I worked up ([source][src]). It correctly does not read the content of the `iframe` when parent has `display: none`, but it incorrectly reads the content for `visibility: hidden`. I'll need to check if this is a known bug...

https://user-images.githubusercontent.com/171986/169715311-20bf45db-e4ae-48ed-a627-2d06749dca26.mov

[tp]: https://7r5p29.csb.app/
[src]: https://codesandbox.io/s/voiceover-hidden-content-test-7r5p29

**Reply by tinymachine** [reply]
This bug of `iframe`s being accessible via VoiceOver in Safari on iOS and macOS even with `visibility: hidden` has indeed [been reported][1] on the WebKit bug tracker (back in 2019!), with no apparent traction so far. (I updated it with [a new screen capture][2].)

[1]: https://bugs.webkit.org/show_bug.cgi?id=201952
[2]: https://bugs.webkit.org/show_bug.cgi?id=201952#c2

**Reply by markteekman** [reply]
Nice find @tinymachine and great reproductions! Damn, I know Safari is still behind on a lot of stuff, but 2019? I don't know what hope there is that it will be fixed soon..

**Reply by tinymachine** [reply]
Thanks! Yeah, agreed - not holding my breath. At least `display: none` is an easy workaround!

**Comment by tinymachine** [comment]
One suggestion might be for your Modal component to toggle `display: none` instead of `visibility: hidden` since that does work correctly on iOS. Any downsides to that?

**Reply by markteekman** [reply]
Well, I had this in the first place, but the problem with `display: none` it that transitions/animations don't work when you toggle to `display: block` (or `flex` or `grid`). That's why I went with `visibility: hidden` 🤔 Did you try to manually override the styles with a `display:none` to see if that works?
